pkg/controllers/framework.go

```diff
@@ -84,55 +118,100 @@ func (km *KindControllerManager) Handle(id string) {
+	reqContext := context.WithValue(ctx, EventID, id)
 
-	log := logger.NewOCMLogger(ctx)
-
-	event, err := km.events.Get(ctx, id)
+	event, svcErr := km.events.Get(reqContext, id)
+	if svcErr != nil {
+		return svcErr
+	}
 
-	if err != nil {
-		log.Error(err.Error())
-		return
+	if event.ReconciledDate != nil {
+		return nil
 	}
 
 	source, found := km.controllers[event.Source]
 	if !found {
-		log.Infof("No controllers found for '%s'\n", event.Source)
-		return
+		logger.Infof("No controllers found for '%s'\n", event.Source)
+		return nil
 	}
 
 	handlerFns, found := source[event.EventType]
 	if !found {
-		log.Infof("No handler functions found for '%s-%s'\n", event.Source, event.EventType)
-		return
+		logger.Infof("No handler functions found for '%s-%s'\n", event.Source, event.EventType)
+		return nil
 	}
 
 	for _, fn := range handlerFns {
-		err := fn(ctx, event.SourceID)
+		err := fn(reqContext, event.SourceID)
 		if err != nil {
-			errStr := fmt.Sprintf("error handing event %s, %s, %s: %s", event.Source, event.EventType, id, err)
-			log.Error(errStr)
-			return
+			return fmt.Errorf("error handing event %s, %s, %s: %s", event.Source, event.EventType, id, err)
 		}
 	}
 
 	// all handlers successfully executed
 	now := time.Now()
 	event.ReconciledDate = &now
-	_, err = km.events.Replace(ctx, event)
+	_, svcErr = km.events.Replace(reqContext, event)
+	return svcErr
+}
+
+func (km *KindControllerManager) runWorker() {
+	// hot loop until we're told to stop. processNextEvent will automatically wait until there's work available, so
+	// we don't worry about secondary waits
+	for km.processNextEvent() {
+	}
+}
+
+// processNextEvent deals with one key off the queue. It returns false when it's time to quit.
+func (km *KindControllerManager) processNextEvent() bool {
+	// pull the next event item from queue.
+	key, quit := km.eventsQueue.Get()
+	if quit {
+		return false
+	}
+	defer km.eventsQueue.Done(key)
+
+	if err := km.handleEvent(key.(string)); err != nil {
+		glog.Errorf("Failed to handle the event %v, %v ", key, err)
+
+		// we failed to handle the event, we should requeue the item to work on later
+		// this method will add a backoff to avoid hotlooping on particular items
+		km.eventsQueue.AddRateLimited(key)
+		return true
+	}
+
+	// we handle the event successfully, tell the queue to stop tracking history for this event
+	km.eventsQueue.Forget(key)
+	return true
+}
+
+func (km *KindControllerManager) processAllEvents() {
+	events, err := km.events.All(context.Background())
 	if err != nil {
-		log.Error(err.Error())
+		// this process is called periodically, so if one error happened, we will wait for the next cycle to handle
+		// this again
+		glog.Errorf("Failed to list events from db, %v", err)
+		return
+	}
+
+	for _, event := range events {
+		// the event has been reconciled, we can remove it from the database safely
+		if event.ReconciledDate != nil {
+			if err := km.events.Delete(context.Background(), event.ID); err != nil {
```

Comment:
we should use a batch delete instead.

Reply:
done, leverage sql batch delete statement